docs/lab/lab04/lab04.md

```diff
@@ -0,0 +1,384 @@
+
+![ClockSignal](images/clock_signal.png)
+
+The most precise type of clock is the crystal oscillator (XOSC). The reason why it is so accurate is because it uses the crystal's natural vibration frequency to create the clock signal. This clock is usually external to the processor itself, but the processor also has an internal clock (ROSC) that is less accurate and that can be used in cases where small variations of clock  pulses are negligeable. When using the USB protocol, for instance, a more stable clock signal is required, therefore the XOSC is necessary.
+The crystal oscillator on the Raspberry Pi Pico board has a frequency of 12MHz. This clock signal is just a reference, and most of the time we need to adjust it to our needs. This is done by either multiplying or dividing the clock, or in other words, elevating or lowering the frequency of the clock. For example, the RP2040 itself runs on a 125MHz clock, so the crystal oscillator frequency of 12MHz is multiplied (this is done using a method called Phase-Locked Loop).
+
+### Counters
+
+A counter is a piece of hardware logic that counts, as its name suggests. Every clock cycle, it increments the value of a register, until it overflows and starts anew. 
+
+:::info
+A regular counter on 8 bits would count up from 0 to 255, then loop back to 0 and continue counting. 
+:::
+
+TODO: insert image from GTKWave
+
+On the RP2040, the counter is associated with 3 registers:
+1. `value` - the current value of the counter
+2. `direction` - whether the counter is counting UP or DOWN
+3. `reset` - if the direction is UP, the value at which the counter resets to 0; if the direction is DOWN, the value at which the counter reset after reaching 0
+
+![Counter](images/counter.svg)
+
+The way the counter works here is that it increments/decrements every clock cycle and checks whether or not it has reached its reset value. If is has, then it resets to its initial value and starts all over again.
+
+The ARM Cortex-M uses the SysTick time counter to keep track of time. This counter is decremented every microsecond, and when it reaches 0, it triggers an exception and then resets.
+
+`SYST_CVR` register = the value of the timer itself
+
+`SYST_RVR` register = the reset value
+
+`SYST_CSR_SET` register:
+	- `ENABLE` field = enable/disable the counter
+	- `TICKINT` field = enable/disable exception on reaching 0
+
+### Timers
+
+Until now, we have been able to blink a led at a certain time interval, by waiting a while between each led toggle. The technique we used so far was asking the processor to skip a clock cycle a number of times, or by calling the processor instruction "nop" (no operation) in a loop. 
+:::info
+This method is not ideal, since the "nop" instruction stalls the processor and wastes valuable time that could otherwise be used to do other things in the meantime. To optimize this, we can use *alarms*.
```

Comment:
```suggestion
This method is not ideal, since the `nop` instruction stalls the processor and wastes valuable time that could otherwise be used to do other things in the meantime. To optimize this, we can use *alarms*.
```